include/nbl/builtin/hlsl/bxdf/ndf.hlsl

```diff
@@ -0,0 +1,339 @@
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_BXDF_NDF_INCLUDED_
+#define _NBL_BUILTIN_HLSL_BXDF_NDF_INCLUDED_
+
+#include "nbl/builtin/hlsl/limits.hlsl"
+#include "nbl/builtin/hlsl/bxdf/common.hlsl"
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace bxdf
+{
+namespace ndf
+{
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+struct SIsotropicParams
+{
+    using this_t = SIsotropicParams<T>;
+
+    static this_t create(T n_or_a2, T NdotH, T NdotH2)    // beckmann, ggx : a2; blinn-phong : n
+    {
+        this_t retval;
+        retval.n_or_a2 = n_or_a2;
+        retval.NdotH = NdotH;
+        retval.NdotH2 = NdotH2;
+        return retval;
+    }
+
+    T n_or_a2;
+    T NdotH;
+    T NdotH2;
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+struct SAnisotropicParams
+{
+    using this_t = SAnisotropicParams<T>;
+
+    static this_t create(T NdotH, T one_minus_NdotH2_rcp, T TdotH2, T BdotH2, T nx, T ny)   // blinn-phong
+    {
+        this_t retval;
+        retval.NdotH = NdotH;
+        retval.one_minus_NdotH2_rcp = one_minus_NdotH2_rcp;
+        retval.TdotH2 = TdotH2;
+        retval.BdotH2 = BdotH2;
+        retval.nx = nx;
+        retval.ny = ny;
+        return retval;
+    }
+
+    static this_t create(T ax, T ay, T ax2, T ay2, T TdotH2, T BdotH2, T NdotH2)    // beckmann, ggx aniso
+    {
+        this_t retval;
+        retval.ax = ax;
+        retval.ax2 = ax2;
+        retval.ay = ay;
+        retval.ay2 = ay2;
+        retval.TdotH2 = TdotH2;
+        retval.BdotH2 = BdotH2;
+        retval.NdotH2 = NdotH2;
+        return retval;
+    }
+
+    static this_t create(T a2, T TdotH, T BdotH, T NdotH)   // ggx burley
+    {
+        this_t retval;
+        retval.ax = a2;
+        retval.TdotH = TdotH;
+        retval.BdotH = BdotH;
+        retval.NdotH = NdotH;
+        return retval;
+    }
+
+    T ax;
+    T ay;
+    T ax2;
+    T ay2;
+    T nx;
+    T ny;
+    T NdotH;
+    T TdotH;
+    T BdotH;
+    T NdotH2;
+    T TdotH2;
+    T BdotH2;
+    T one_minus_NdotH2_rcp;
+};
+
+
+template<typename T NBL_PRIMARY_REQUIRES(is_scalar_v<T>)
+struct BlinnPhong
+{
+    using scalar_type = T;
+
+    // blinn-phong
+    scalar_type operator()(SIsotropicParams<scalar_type> params)
+    {
+        // n is shininess exponent in original paper
+        return isinf<scalar_type>(params.n_or_a2) ? numeric_limits<scalar_type>::infinity : numbers::inv_pi<scalar_type> * 0.5 * (params.n_or_a2 + 2.0) * pow<scalar_type>(params.NdotH, params.n_or_a2);
+    }
+
+    //ashikhmin-shirley ndf
+    scalar_type operator()(SAnisotropicParams<scalar_type> params)
+    {
+        scalar_type n = (params.TdotH2 * params.ny + params.BdotH2 * params.nx) * params.one_minus_NdotH2_rcp;
+        return (isinf<scalar_type>(params.nx) || isinf<scalar_type>(params.ny)) ?  numeric_limits<scalar_type>::infinity : 
```

Comment:
its enough to check that `n` is "large enough" e.g. >254 because `pow` with exponent higher than 254 is bound to make a lowest non denormalized number into an inf